plugins/user-settings/src/components/useUserProfileInfo.ts

```diff
@@ -14,13 +14,39 @@
  * limitations under the License.
  */
 
-import { useApi, identityApiRef } from '@backstage/core-plugin-api';
+import {
+  alertApiRef,
+  identityApiRef,
+  useApi,
+} from '@backstage/core-plugin-api';
+import { useEffect } from 'react';
+import { useAsync } from 'react-use';
 
 export const useUserProfile = () => {
   const identityApi = useApi(identityApiRef);
-  const userId = identityApi.getUserId();
-  const profile = identityApi.getProfile();
-  const displayName = profile.displayName ?? userId;
+  const alertApi = useApi(alertApiRef);
 
-  return { profile, displayName };
+  const { value, loading, error } = useAsync(async () => {
+    return {
+      profile: await identityApi.getProfileInfo(),
+      identity: await identityApi.getBackstageIdentity(),
+    };
+  }, []);
+
+  useEffect(() => {
+    if (error) {
+      alertApi.post({
+        message: `Failed to load user identity: ${error}`,
+        severity: "error",
+      });
+    }
+  }, [error, alertApi]);
+
+  return {
+    profile: loading ? {} : value!.profile,
```

Comment:
Oh this was actually a nice catch of yours - the typescript is actually right here, we can't be sure at this point that it has a value - because we aren't exiting when `error` is true. We have to address that.

maybe something like

```ts
if (loading || error) {
  return {
    profile: {},
    displayName: '',
    loading,
  };
};

return {
  profile: value.profile, // at this point tsc will know that it actually isn't undefined!
  displayName: value.profile.displayName ?? value.identity.userEntityRef,
  loading: false,
};
```

that actually even turned out to be cleaner.

Reply:
Ah of course. I added the check and reutrn on `loading || error` but tsc was still complaining about using `value` :disappointed:. But now it should actually be safe to use `!`.